pwiz_tools/Skyline/Model/Results/RemoteApi/Ardia/ArdiaUrl.cs

```diff
@@ -0,0 +1,267 @@
+
+            public override void SetLength(long value)
+            {
+                _input.SetLength(value);
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
+            {
+                int n = _input.Read(buffer, offset, count);
+                _position += n;
+                UpdateProgress?.Invoke(this, new ProgressEventArgs((1.0f * _position) / _length));
+                return n;
+            }
+
+            public override void Write(byte[] buffer, int offset, int count)
+            {
+                throw new NotImplementedException();
+            }
+
+            public override bool CanRead => true;
+            public override bool CanSeek => false;
+            public override bool CanWrite => false;
+            public override long Length => _length;
+            public override long Position
+            {
+                get { return _position; }
+                set { throw new NotImplementedException(); }
+            }
+        }
+
+        public override MsDataFileImpl OpenMsDataFile(OpenMsDataFileParams openMsDataFileParams)
+        {
+            var rawFilepath = Path.Combine(openMsDataFileParams.DownloadPath, RawName);
+            if (File.Exists(rawFilepath))
+                return openMsDataFileParams.OpenLocalFile(rawFilepath, 0, LockMassParameters);
+
+            var account = FindMatchingAccount(Settings.Default.RemoteAccountList) as ArdiaAccount;
+            if (account == null)
+            {
+                throw new RemoteServerException(string.Format(ArdiaResources.UnifiUrl_OpenMsDataFile_Cannot_find_account_for_username__0__and_server__1__, 
+                    Username, ServerUrl));
+            }
+
+            if (StorageId.IsNullOrEmpty())
+                throw new InvalidDataException(ArdiaResources.ArdiaUrl_OpenMsDataFile_cannot_open_an_ArdiaUrl_because_it_is_not_a_RAW_file_URL_with_a_StorageId);
+
+            var p = openMsDataFileParams;
+            using var client = account.GetAuthenticatedHttpClient();
+            string presignedUrl = GetPresignedUrl(client, StorageId);
+
+            p.ProgressStatus = p.ProgressStatus.ChangeSegments(p.ProgressStatus.Segment, Math.Max(1, p.ProgressStatus.SegmentCount) + 1);
+            p.ProgressMonitor.UpdateProgress(p.ProgressStatus);
+
+            var response = client.GetAsync(presignedUrl, HttpCompletionOption.ResponseHeadersRead, openMsDataFileParams.CancellationToken).Result;
+            response.EnsureSuccessStatusCode();
+            var responseStream = response.Content.ReadAsStreamAsync().Result;
+            using (var progressStream = new ProgressStream(responseStream, RawSize ?? 1))
+            using (var fileStream = new FileStream(rawFilepath, FileMode.CreateNew))
+            {
+                void ProgressStream_OnUpdateProgress(object sender, ProgressEventArgs e)
+                {
+                    var newPercentComplete = (int) Math.Round(e.Progress * 100);
```

Comment:
ProgressStatus objects do something a little weird when you set their progress to 100, so you might want to change this to:
`var newPercentComplete = Math.Min(99, (int) Math.Round(e.Progress * 100));`

The specific weird thing that they do is that if they have multiple segments, the first time you set the progress to 100, "ProgressStatus.ChangePercentComplete" does the following:
`// Turn off progress zooming, if the end has been reached`
and the second time you call ChangePercentComplete(100) it sets "State" to ProgressState.complete

Once the progress state has been set to "Complete", "MultiProgressStatus.ChangeStatus" will stop noticing changes on it because:
`// Avoid overwriting a final status with a non-final status`

Reply:
Is there a good way to indicate that the importing stage is actually downloading? There doesn't seem to be an individual text status for replicates in the AllChromatogramsGraph dialog.

Reply:
I can't think of a good way to do something like that.

A different approach that we could take could be to make it so that Skyline does not even start trying to extract chromatograms until all of the Ardia files have been downloaded.
That is, we would implement a BackgroundLoader ("ArdiaDownloadManager") that looks at the MsDataFileUri's in the document and figures out whether any of them are ArdiaUrl objects, and, if so, starts downloading them.
We would change "ChromatogramManager.IsReadyToLoad" so that it returns false if the ArdiaDownloadManager says it's not finished.
In this way, the experience would be similar to what the user gets when their are retention time alignments that need to be performed by RetentionTimeManager, and DocumentRetentionTimes.IsNotLoadedExplained returns something indicating that there is still work to be done.
The ArdiaDownloadManager could provide progress that would show up in the Status Bar at the bottom of the Skyline window.

Doing it that way would probably be a worse experience because no chromatogram extraction would happen until all of the files had been downloaded. Also, it's hard to see how to make it work with the "delete after chromatogram extraction" checkbox. But, it's something to keep in mind as a possibility.